File: src/integrations/amapGeocoder.ts
export type AmapConvertedLocation = {
  provider: "amap";
  coordSystem: "gcj02";
  longitude: number;
  latitude: number;
  rawLocation: string;
};

export type AmapParsedRegeo = {
  provider: "amap";
  formattedAddress: string;
  address: {
    province?: string;
    city?: string;
    district?: string;
    township?: string;
  };
  pois: Array<{
    name: string;
    type?: string;
    distanceMeters?: number;
    direction?: string;
  }>;
  roads: Array<{
    name: string;
    distanceMeters?: number;
    direction?: string;
  }>;
};

export function buildAmapConvertUrl(input: {
  key: string;
  longitude: number;
  latitude: number;
}) {
  const url = new URL("https://restapi.amap.com/v3/assistant/coordinate/convert");
  url.searchParams.set("key", input.key);
  url.searchParams.set("locations", `${input.longitude},${input.latitude}`);
  url.searchParams.set("coordsys", "gps");
  return url;
}

export function parseAmapConvert(response: {
  status?: string;
  info?: string;
  infocode?: string;
  locations?: string;
}): AmapConvertedLocation {
  if (response.status !== "1" || !response.locations) {
    throw new Error(`Amap coordinate conversion failed: ${response.info ?? response.infocode ?? "unknown error"}`);
  }

  const [longitude, latitude] = response.locations.split(",").map((value) => Number(value));
  if (!Number.isFinite(longitude) || !Number.isFinite(latitude)) {
    throw new Error(`Amap coordinate conversion returned invalid location: ${response.locations}`);
  }

  return {
    provider: "amap",
    coordSystem: "gcj02",
    longitude,
    latitude,
    rawLocation: response.locations
  };
}

export function buildAmapRegeoUrl(input: {
  key: string;
  amapLocation: string;
  radius?: number;
}) {
  const url = new URL("https://restapi.amap.com/v3/geocode/regeo");
  url.searchParams.set("key", input.key);
  url.searchParams.set("location", input.amapLocation);
  url.searchParams.set("radius", String(input.radius ?? 1000));
  url.searchParams.set("extensions", "all");
  return url;
}

const optionalNumber = (value: unknown) => {
  const numberValue = Number(value);
  return Number.isFinite(numberValue) ? numberValue : undefined;
};

export function parseAmapRegeo(response: {
  status?: string;
  info?: string;
  infocode?: string;
  regeocode?: {
    formatted_address?: string;
    addressComponent?: {
      province?: string;
      city?: string;
      district?: string;
      township?: string;
    };
    pois?: Array<{
      name?: string;
      type?: string;
      distance?: string;
      direction?: string;
    }>;
    roads?: Array<{
      name?: string;
      distance?: string;
      direction?: string;
    }>;
  };
}): AmapParsedRegeo {
  if (response.status !== "1" || !response.regeocode) {
    throw new Error(`Amap reverse geocode failed: ${response.info ?? response.infocode ?? "unknown error"}`);
  }

  const regeocode = response.regeocode;

  return {
    provider: "amap",
    formattedAddress: regeocode.formatted_address ?? "",
    address: {
      province: regeocode.addressComponent?.province,
      city: regeocode.addressComponent?.city,
      district: regeocode.addressComponent?.district,
      township: regeocode.addressComponent?.township
    },
    pois: (regeocode.pois ?? [])
      .filter((poi) => poi.name)
      .map((poi) => ({
        name: poi.name as string,
        type: poi.type,
        distanceMeters: optionalNumber(poi.distance),
        direction: poi.direction
      })),
    roads: (regeocode.roads ?? [])
      .filter((road) => road.name)
      .map((road) => ({
        name: road.name as string,
        distanceMeters: optionalNumber(road.distance),
        direction: road.direction
      }))
  };
}

File: native/MemoryMapNative/Sources/MemoryMapNativeCore/AmapGeocoder.swift
import Foundation

public struct AmapConvertedLocation: Codable, Equatable, Sendable {
    public let provider: String
    public let coordSystem: String
    public let longitude: Double
    public let latitude: Double
    public let rawLocation: String
}

public struct AmapAddress: Codable, Equatable, Sendable {
    public let province: String?
    public let city: String?
    public let district: String?
    public let township: String?
}

public struct AmapPoi: Codable, Equatable, Sendable {
    public let name: String
    public let type: String?
    public let distanceMeters: Double?
    public let direction: String?
}

public struct AmapRoad: Codable, Equatable, Sendable {
    public let name: String
    public let distanceMeters: Double?
    public let direction: String?
}

public struct AmapParsedRegeo: Codable, Equatable, Sendable {
    public let provider: String
    public let formattedAddress: String
    public let address: AmapAddress
    public let pois: [AmapPoi]
    public let roads: [AmapRoad]
}

public enum AmapGeocoder {
    public static func buildConvertURL(key: String, longitude: Double, latitude: Double) throws -> URL {
        var components = URLComponents(string: "https://restapi.amap.com/v3/assistant/coordinate/convert")!
        components.queryItems = [
            URLQueryItem(name: "key", value: key),
            URLQueryItem(name: "locations", value: "\(longitude),\(latitude)"),
            URLQueryItem(name: "coordsys", value: "gps")
        ]
        guard let url = components.url else { throw AmapError.invalidURL }
        return url
    }

    public static func buildRegeoURL(key: String, amapLocation: String, radius: Int = 1000) throws -> URL {
        var components = URLComponents(string: "https://restapi.amap.com/v3/geocode/regeo")!
        components.queryItems = [
            URLQueryItem(name: "key", value: key),
            URLQueryItem(name: "location", value: amapLocation),
            URLQueryItem(name: "radius", value: "\(radius)"),
            URLQueryItem(name: "extensions", value: "all")
        ]
        guard let url = components.url else { throw AmapError.invalidURL }
        return url
    }

    public static func parseConvert(_ data: Data) throws -> AmapConvertedLocation {
        let response = try JSONDecoder().decode(ConvertResponse.self, from: data)
        guard response.status == "1", let locations = response.locations else {
            throw AmapError.service(response.info ?? response.infocode ?? "unknown error")
        }
        let parts = locations.split(separator: ",").compactMap { Double($0) }
        guard parts.count == 2 else { throw AmapError.service("invalid location: \(locations)") }
        return AmapConvertedLocation(provider: "amap", coordSystem: "gcj02", longitude: parts[0], latitude: parts[1], rawLocation: locations)
    }

    public static func parseRegeo(_ data: Data) throws -> AmapParsedRegeo {
        let response = try JSONDecoder().decode(RegeoResponse.self, from: data)
        guard response.status == "1", let regeocode = response.regeocode else {
            throw AmapError.service(response.info ?? response.infocode ?? "unknown error")
        }
        return AmapParsedRegeo(
            provider: "amap",
            formattedAddress: regeocode.formattedAddress ?? "",
            address: AmapAddress(
                province: regeocode.addressComponent?.province,
                city: regeocode.addressComponent?.city,
                district: regeocode.addressComponent?.district,
                township: regeocode.addressComponent?.township
            ),
            pois: (regeocode.pois ?? []).compactMap { poi in
                guard let name = poi.name else { return nil }
                return AmapPoi(name: name, type: poi.type, distanceMeters: Double(poi.distance ?? ""), direction: poi.direction)
            },
            roads: (regeocode.roads ?? []).compactMap { road in
                guard let name = road.name else { return nil }
                return AmapRoad(name: name, distanceMeters: Double(road.distance ?? ""), direction: road.direction)
            }
        )
    }
}

public enum AmapError: Error, Equatable {
    case invalidURL
    case service(String)
}

private struct ConvertResponse: Decodable {
    let status: String?
    let info: String?
    let infocode: String?
    let locations: String?
}

private struct RegeoResponse: Decodable {
    let status: String?
    let info: String?
    let infocode: String?
    let regeocode: Regeo?
}

private struct Regeo: Decodable {
    let formattedAddress: String?
    let addressComponent: AddressComponent?
    let pois: [Poi]?
    let roads: [Road]?

    enum CodingKeys: String, CodingKey {
        case formattedAddress = "formatted_address"
        case addressComponent
        case pois
        case roads
    }
}

private struct AddressComponent: Decodable {
    let province: String?
    let city: String?
    let district: String?
    let township: String?
}

private struct Poi: Decodable {
    let name: String?
    let type: String?
    let distance: String?
    let direction: String?
}

private struct Road: Decodable {
    let name: String?
    let distance: String?
    let direction: String?
}
